core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/AdaptiveReorderPartitionedJoin.java

```diff
@@ -0,0 +1,258 @@
+import static io.trino.sql.planner.plan.Patterns.join;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Flip sides for partitioned join based on data size stats. This rule is only used during Adaptive Planning in FTE.
+ * From:
+ * <pre>
+ *    Join (PARTITIONED)
+ *      - left side (partitioned)
+ *      - Partial Aggregation (optional)
+ *          - Exchange (LOCAL)
+ *              - right side (partitioned)
+ * </pre>
+ * TO:
+ * <pre>
+ *    Join (PARTITIONED)
+ *      - Partial Aggregation (optional)
+ *          - right side (partitioned)
+ *      - Exchange (LOCAL) (if needed, derived using source stream properties)
+ *          - left side (partitioned)
+ * </pre>
+ */
+public class AdaptiveReorderPartitionedJoin
+        implements Rule<JoinNode>
+{
+    private static final Capture<ExchangeNode> LOCAL_EXCHANGE_NODE = Capture.newCapture();
+    private static final Pattern<JoinNode> PATTERN = join()
+            .matching(AdaptiveReorderPartitionedJoin::isPartitionedJoinWithNoHashSymbols)
+            .or(
+                    // In case partial aggregation is missing
+                    prev -> prev.with(right().matching(
+                            exchange().matching(exchangeNode -> exchangeNode.getScope().equals(LOCAL)
+                                            // Skip when exchange is gather
+                                            && !exchangeNode.getType().equals(GATHER))
+                                    .capturedAs(LOCAL_EXCHANGE_NODE))),
+                    // In case partial aggregation is present
+                    prev -> prev.with(right().matching(
+                            aggregation().matching(node -> node.getStep() == PARTIAL)
+                                    .with(source().matching(
+                                            exchange().matching(exchangeNode -> exchangeNode.getScope().equals(LOCAL)
+                                                            // Skip when exchange is gather
+                                                            && !exchangeNode.getType().equals(GATHER))
+                                                    .capturedAs(LOCAL_EXCHANGE_NODE))))));
+
+    // Use a smaller threshold to change the order since the cost of changing the order is lower here. Additionally,
+    // the data size stats are more accurate compared to static planning since they are collected at run time from the
+    // stage execution.
+    private static final double SIZE_DIFFERENCE_THRESHOLD = 1.5;
+
+    // With speculative execution, reordering small joins might cause unnecessary restarts of the
+    // join stage and lead to performance degradation. Hence, we only reorder if the size of the right side is
+    // above this threshold.
+    // TODO: Consider the cost of restarting the stage as part of adaptive planning.
+    private static final double MIN_SIZE_THRESHOLD = DataSize.of(5, GIGABYTE).toBytes();
```

Comment:
can we make those configurable (or at least threshold) so we can do some experiments what works best?